drgn/helpers/linux/idr.py

```diff
@@ -7,21 +7,101 @@
 
 The ``drgn.helpers.linux.idr`` module provides helpers for working with the IDR
 data structure in :linux:`include/linux/idr.h`. An IDR provides a mapping from
-an ID to a pointer. This currently only supports Linux v4.11+; before this,
-IDRs were not based on radix trees.
+an ID to a pointer.
 """
 
 from typing import Iterator, Tuple
 
-from _drgn import _linux_helper_idr_find as idr_find
-from drgn import Object
+from _drgn import _linux_helper_idr_find as idr_find_in_radix_tree
+from drgn import NULL, Object, Program, sizeof
 from drgn.helpers.linux.radixtree import radix_tree_for_each
 
 __all__ = (
     "idr_find",
     "idr_for_each",
 )
 
+IDR_BITS = 8
+IDR_MASK = (1 << IDR_BITS) - 1
+
+
+def _max_idr_shift(prog: Program) -> int:
+    return sizeof(prog.type("int")) * 8 - 1
+
+
+def _max_idr_level(prog: Program) -> int:
+    return (_max_idr_shift(prog) + IDR_BITS - 1) // IDR_BITS
+
+
+def _round_up(number: int, multiple: int) -> int:
+    return -((-number) // multiple) * multiple
+
+
+def _idr_max(prog: Program, layers: int) -> int:
+    bits = min(layers * IDR_BITS, _max_idr_shift(prog))
+    return (1 << bits) - 1
+
+
+def _fls(num: int) -> int:
+    num_bits = 32
+    if num == 0:
+        return 0
+
+    for i in range(num_bits):
+        if (num & 1) == 1:
+            pos = i
+
+        num = num >> 1
+
+    return pos + 1
+
+
+def idr_find(idr: Object, idval: int) -> Object:
```

Comment:
To match the original signature, `idval` should be named `id`, and it should be an `IntegerLike` so that `id` can be given as a drgn object:

```suggestion
def idr_find(idr: Object, id: IntegerLike) -> Object:
```

Then below to convert it to an `int`, you can do:

```python3
    id = operator.index(id)
```

You'll of course need:

```python3
import operator
from drgn import IntegerLike
```

Reply:
Sure. I have incorporated these changes